File: Poly.java
package poly;

import java.util.Scanner;
import java.util.Vector;


class Array{
    private int initial;
    private long[] nums = new long[Poly.len];

    public Array(){

    }
    public Array(int size){
        initial = size;
    }
    public void setNum(long num, int degree){
        nums[degree] = num + nums[degree];
    }
    public long getNum(int degree){
        return nums[degree];
    }
}

public class Poly {
    public Poly(){

    }
    public Poly(String input){

    }
    public static int len = 10000;
    private Array[] array = new Array[len];


    public Array result(String[] subString){
        Array result = new Array(len);
        return result;
    }

    public Array add(Array x, Array y){//加法
        int i;
        i = 0;
        Array result = new Array();
        for(i = 0; i < len; i++){
            result.setNum(x.getNum(i) + y.getNum(i), i);
        }
        return result;
    }
    public Array sub(Array x, Array y){//减法
        int i;
        i = 0;
        Array result = new Array();
        for(i = 0; i < len; i++){
            result.setNum(x.getNum(i) - y.getNum(i), i);
        }
        return result;
    }

    public boolean geshi(String subString){//正则表达式
        if(!subString.matches(
                "^[+-]?(\\{(\\(([+-]?[0-9]{1,8},[+]?[0-9]{1,6})?\\),)*"
                        + "\\(([+-]?[0-9]{1,8},[+]?[0-9]{1,6})?\\)\\}[+-])*"
                        + "\\{(\\(([+-]?[0-9]{1,8},[+]?[0-9]{1,6})?\\),)*"
                        + "\\(([+-]?[0-9]{1,8},[+]?[0-9]{1,6})?\\)\\}$"))
            return false;
        else
            return true;
    }

    public long[] getPoly(String subString){
        int i, j;
        i = 0;
        j = 0;
        long[] nums = new long[len];
        String[] numString = new String[len];
        numString = subString.split("[{(),}]");
        for(i = 0, j = 0; i < numString.length; i++){
            if(!numString[i].equals("")){
                nums[j++] = Integer.parseInt(numString[i]);
            }
        }
        return nums;
    }

    public Array getArray(long[] num){
        int i;
        i = 0;
        Array array = new Array(len);
        for(i = 0; i < num.length; i = i + 2){
            array.setNum(num[i], (int)num[i+1]);
        }
        return array;
    }

    public String operator(String input){
        int i;
        String op = new String();
        String[] axx = new String[len];
        axx = input.split("\\{(\\([+-]?[0-9]{1,8}\\,[+]?[0-9]{1,6}\\),)*\\([+-]?[0-9]{1,8}\\,[+]?[0-9]{1,6}\\)\\}");
        op = op + "+";
        for(i = 0; i < axx.length; i++){
            if(!axx[i].equals("")){
                op  = op + axx[i];
            }
        }
        return op;
    }

    public void print(Array result){
        int i;
        int judge;
        i = 0;
        judge = 0;
        System.out.print("{");
        for(i=0 ; i < len; i++){
            if(result.getNum(i) != 0){
                //System.out.print(result.getNum(i));
                System.out.print("(" + result.getNum(i) + "," + i + ")" + ",");
                judge = 1;
            }
        }
        if(judge == 1){
            //System.out.println("}");
            System.out.println("\b}");
        }
        else
            System.out.println("()}");
    }

    public void spacedel(String input){//删除空格
        String[] subString = new String[len];
        subString = input.split("\\}[+-]\\{");
		/*  int l,i;
     l=String.lenth;
     for(i=0;i<l;i++)
     {for(p=string;p;p++)
     if(p==' ')
     strcpy(p,p+1);}*/
    }

    public static void main(String argv[]){
        String init = "{(0,0)}";
        int i;
        i = 0;
        while(true){//try catch
            try{
                System.out.println("请输入要计算的多项式：");
                Poly poly = new Poly();
                Scanner scanner = new Scanner(System.in);
                String input = scanner.nextLine();

                if(input.startsWith("+") || input.startsWith("-")){
                    input = init + input;
                }
                input = input.replace(" ", "");//
              //  input =input.replace("  ","")//tab
                String[] subString = new String[len];
                subString = input.split("\\}[+-]\\{");
                Array array = new Array();
                Array result = new Array();
                String op = poly.operator(input);
                if(poly.geshi(input) == true){
                    long [] nums = new long[len];
                    for(i = 0; i < subString.length; i++){
                        nums = poly.getPoly(subString[i]);
                        array = poly.getArray(nums);

                        if(op.charAt(i) == '-'){
                            result = poly.sub(result, array);
                        }
                        else if(op.charAt(i) == '+'){
                            result = poly.add(result, array);
                        }
                    }
                    System.out.println("计算结果为：");
                    poly.print(result);
                }
                else
                    System.out.println("输入格式错误");
            }
            catch(Throwable t){
                System.out.println("error：" + t);
                continue;
            }
            finally{

            }
        }
    }

}

File: poly.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

void spacedel(char string[]){//É¾³ý¿Õ¸ñ
     char *p;
     int l,i;
     l=strlen(string);
     for(i=0;i<l;i++)
     {for(p=string;*p;p++)
     if(*p==' ')
     strcpy(p,p+1);}
}

int add(){

}

int sub(){

}
/*typedef struct Poly{
    int coeff[];
    int degree[];
}*/
int main(){

	int i=0,j=-1,op=1;
	int coeff[10000]={0},degree[10000]={0},ci=0,di=0;//coeffÊÇÏµÊý£¬degreeÊÇÃÝ
	int p1=0,p2=0,t=1;  //ÓÃÓÚÌáÈ¡ÏµÊýºÍÃÝ
	int resultc[1000]={0},resultd[1000]={0},x=0,sum=0;
	char s[10000]={0},stack[10000]={0};
	gets(s);
	spacedel(s);

	for(i=0;s[i]!='\0';i++){
			if(s[i]!='{'){//²»ÒÔ{¿ªÍ·±¨´í
				printf("输入错误");
				return 0;
			}
			break;
	}

	for(i=0;s[i]!='\0';i++){
            if(s[i]=='+'){
                if(j==-1){
                    op=1;
                    if(s[i+1]!='{'){//
                        printf("输入错误");
                        return 0;
                    }
                }
                else{
                    stack[++j]=s[i];
                }

		}
		else if(s[i]=='-'){
			if(j==-1){
				op=-1;
				if(s[i+1]!='{'){
					printf("输入错误");
					return 0;
				}
			}
			else{
				stack[++j]=s[i];
			}
		}

		else{
			if(s[i]=='}'){
				if(stack[j]='{' && j==0){
					j--;
				}
				else{
					printf("输入错误");
					return 0;
				}
			}
			else if(s[i]!=')'){
				stack[++j]=s[i];
			}

			else if(s[i]==')'){
				if(s[i+1]==','){
					i++;
					if(s[i+1]!='('){
						printf("输入错误");
						return 0;
					}
				}
				else if(s[i+1]=='}'){
                        printf("输入错误");
						return 0;
				}
				else{
					printf("输入错误");
					return 0;
				}
				if(stack[j]=='('){
					printf("输入错误");
					return 0;
				}
				while(stack[j]!='(' && j>0){
					if(stack[j]==','){
						printf("输入错误");
						return 0;
						}
					else if(isdigit(stack[j])){
						while(isdigit(stack[j])){
							p2=p2+(stack[j]-'0')*t;
							t=t*10;
							j--;
						}
						t=1;
						if(stack[j]=='-'){
							p2=-p2;
							j--;
						}
						else if(stack[j]=='+'){
							j--;
						}
						if(stack[j]==','){
							j--;
							if(isdigit(stack[j])){
								while(isdigit(stack[j])){
									p1=p1+(stack[j]-'0')*t;
									t=t*10;
									j--;
								}
								t=1;
								if(stack[j]=='-'){
									p1=-p1;
									j--;
								}
								else if(stack[j]=='+'){
									j--;
								}
							}
							else{
								printf("输入错误");
								return 0;
							}
						}
						else{
							printf("输入错误");
							return 0;
						}
						if(stack[j]!='(' || j<1){
							printf("输入错误");
							return 0;
						}
						coeff[ci++]=p1*op;
						degree[di++]=p2;
						p1=0;
						p2=0;
					}

					else{
						printf("输入错误");
						return 0;
					}
				}

				if(stack[j]=='('){
					j--;
				}
			}
		}
	}
	if(j!=-1){
		printf("输入错误");
		return 0;
	}


	for(i=0;i<ci;i++){
		if(coeff[i]==0){
			continue;
		}
		else{
			sum=coeff[i];
			for(j=i+1;j<di;j++){
				if(degree[i]==degree[j]){
					sum+=coeff[j];
					coeff[j]=0;
				}
			}
			if(sum==0){

			}
			else{
				resultc[x]=sum;
				resultd[x++]=degree[i];
				sum=0;
			}
		}
	}

	//Êä³ö
	/*printf("计算结果系数为：");
	for(i=0;i<x;i++){
		printf("%d,",resultc[i]);
	}
	printf("\n计算结果次数为:");
	for(i=0;i<x;i++){
		printf("%d,",resultd[i]);
	}*/

	for(i=0;i<x;i++){
        printf("%d %d\n",resultc[i],resultd[i]);
	}


}

